big refactoring of Trainer

joelgrus opened the pull request:
* broke `Trainer.train()` up into a lot of different functions
* got rid of all the `None` checks by writing wrapper functions that no-op in those cases
* created `TensorboardWriter` abstraction for the same reason

let me know if you don't like any of the design choices here

Fixes #204

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -125,18 +143,16 @@ def __init__(self,
+        if serialization_dir is not None:
+            train_log = SummaryWriter(os.path.join(serialization_dir, "log", "train"))
+            validation_log = SummaryWriter(os.path.join(serialization_dir, "log", "validation"))
+            self._writer = TensorboardWriter(train_log, validation_log)
```
I think this name would be better if it had a clear reference to tensorboard in it.  E.g., I'd prefer `_tensorboard` to `_writer`.  `_writer` could be writing lots of different things, but `_tensorboard` is pretty obvious.

joelgrus replied:
done

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -125,18 +143,16 @@ def __init__(self,
+
+    def _enable_gradient_clipping(self) -> None:
         if self._grad_clipping is not None:
             # Pylint is unable to tell that we're in the case that _glad_clipping is not None...
```
Just noticed this comment says "_glad_clipping" instead of "_grad_clipping"...  Care to fix my typo?

joelgrus replied:
done

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+    def _compute_loss(self, batch: torch.Tensor, for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batch and returns the ``loss`` value in the result.
+        Also performs backpropagation if ``for_training`` is ``True``.
```
If this is going to perform backprop, I think it should say so in its name: `_compute_and_backprop_loss`.  It's pretty unintuitive to expect a `_compute_loss` function to also do backprop.

matt-gardner replied:
Actually, why put backprop in this function at all?  Why not just leave it in `_train_epoch`?  Seems cleaner.

joelgrus replied:
done

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+            train_generator_tqdm.set_description(description)
+
+            # Log parameter values to Tensorboard
+            batch_num_total = num_training_batches * epoch + batch_num
```
Can this block be pulled out into its own function, too?

matt-gardner replied:
Also, doesn't look like all of the metrics are getting logged - shouldn't that be happening?

joelgrus replied:
I can add that

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+            # .cpu() is a no-op if you aren't using GPUs.
+            train_loss += loss.data.cpu().numpy()
+
+            # Rescale gradients
```
Most of these comments are just annotating a single line now, where the function being called already has a descriptive name.  I think you can remove these comments.

joelgrus replied:
done

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+                self._last_log = time.time()
+
+        # Return the metrics (and reset them)
+        return self._get_metrics_for_batch(train_loss, batch_num, reset=True)
```
This is getting called twice, looks like.

joelgrus replied:
are you sure? the previous call is per-batch inside the loop, this is at the end of the loop to reset the metrics. or am I missing one?

matt-gardner replied:
Ah, yeah, you're right.  I think the name referring to `for_batch` threw me off.  Not sure how to make the name better, though...

joelgrus replied:
I renamed to make this clearer

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+            val_generator_tqdm.set_description(description)
+
+            # Log progress in the no-tqdm case
+            if self._no_tqdm and time.time() - self._last_log > self._log_interval:
```
I think this was lost in `_train_epoch`.

matt-gardner replied:
You could take this block and the two lines above (`description = ...; ...tqdm.set_description()`) and put them into a function that gets called by both train and validation, too, I think.

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+        # Log parameter values to Tensorboard
+        val_metrics = self._get_metrics_for_batch(val_loss, batch_num, reset=True)
+        message_template = "Training %s : %3f    Validation %s : %3f "
+        for name, value in metrics.items():
```
I'd _probably_ be fine with just having the `_train_epoch` function log the train metrics, and not have to pass them to this function.  If we do pass them here, I'd definitely at least call them `train_metrics` instead of just `metrics`.

joelgrus replied:
I changed the name

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+            self._writer.add_validation_scalar(name, val_metrics[name], epoch)
+
+        # Check validation metric for early stopping
+        this_epoch_val_metric = val_metrics[self._validation_metric]
```
Can this be its own function?

joelgrus replied:
done

matt-gardner commented on allennlp/training/trainer.py:
```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
+            metrics = self._train_epoch(epoch)
+
+            # Validate and move on to the next epoch
+            if self._should_stop(epoch, metrics, validation_metric_per_epoch):
```
I feel like this one went a bit too far - I'd probably put an explicit check for validation data here, and a call to `validate` if so, otherwise just save a checkpoint, then put the learning rate scheduler update stuff after that.  Having the `next_epoch{_no}_validation` seems a little bit much to me.

joelgrus commented:
ok, I think this is ready for another look. I addressed most of the comments, there were a couple of "maybe factor this out" that I couldn't figure out a non-messy way to do, but I think it looks pretty good now.

DeNeutoy commented on allennlp/training/trainer.py:
```diff
@@ -31,6 +31,24 @@
+        if self._train_log is not None:
+            self._train_log.add_scalar(name, value, global_step)
+
+    def add_validation_scalar(self, name: str, value: float, global_step=None) -> None:
```
Why is global_step optional?

joelgrus replied:
I'm sure I must have had a good reason for it, but I couldn't tell you what it is now.

joelgrus replied:
I'll make it non-optional.
